src/zkevm_specs/evm_circuit/execution/error_oog_memory_copy.py

```diff
@@ -0,0 +1,66 @@
+from zkevm_specs.evm_circuit.table import CallContextFieldTag
+from zkevm_specs.util.param import (
+    GAS_COST_ACCOUNT_COLD_ACCESS,
+    GAS_COST_FASTEST,
+    GAS_COST_WARM_ACCESS,
+    N_BYTES_MEMORY_ADDRESS,
+)
+from ...util import FQ
+from ..instruction import Instruction
+from ..opcode import Opcode
+from ...util import N_BYTES_GAS
+
+
+def error_oog_memory_copy(instruction: Instruction):
+    # retrieve op code associated to oog constant error
+    opcode = instruction.opcode_lookup(True)
+    (
+        is_calldata_copy,
+        is_code_copy,
+        is_ext_code_copy,
+        is_return_data_copy,
+    ) = instruction.multiple_select(
+        opcode, (Opcode.CALLDATACOPY, Opcode.CODECOPY, Opcode.EXTCODECOPY, Opcode.RETURNDATACOPY)
+    )
+
+    # Constrain opcode must be one of CALLDATACOPY, CODECOPY, EXTCODECOPY or RETURNDATACOPY.
+    instruction.constrain_equal(
+        is_calldata_copy + is_code_copy + is_ext_code_copy + is_return_data_copy, FQ(1)
+    )
+
+    # EXTCODECOPY has one more extra stack pop for external address
+    if is_ext_code_copy == FQ(1):
+        external_address = instruction.stack_pop()
+    memory_offset_word = instruction.stack_pop()
+    instruction.stack_pop()  # calldata offset, we don't need it here.
+    copy_size_word = instruction.stack_pop()
+
+    # Get constant gas cost
+    if is_ext_code_copy == FQ(1):
+        address = instruction.word_to_fq(external_address, N_BYTES_MEMORY_ADDRESS)
+        tx_id = instruction.call_context_lookup(CallContextFieldTag.TxId)
+        is_warm = instruction.add_account_to_access_list(
+            tx_id, address, instruction.reversion_info()
+        )
+        if is_warm == FQ(1):
+            constant_gas = GAS_COST_WARM_ACCESS
+        else:
+            constant_gas = GAS_COST_ACCOUNT_COLD_ACCESS
+    else:
+        constant_gas = GAS_COST_FASTEST
+
+    # get dynamic gas cost which includes memory expansion and memory copy
+    memory_offset = instruction.word_to_fq(memory_offset_word, N_BYTES_MEMORY_ADDRESS)
+    copy_size = instruction.word_to_fq(copy_size_word, N_BYTES_MEMORY_ADDRESS)
+    _, memory_expansion_gas_cost = instruction.memory_expansion_dynamic_length(
```

Comment:
I checked the computation of `memory_expansion_gas_cost`, it does not seem to be the same as in the md file. Is there an error in either of them?

Reply:
Could you help point out which part is different?

You could also check the calculation of `memory_expansion_gas_cost` [here](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/specs/opcode/F1CALL_F2CALLCODE_F4DELEGATECALL_FASTATICCALL.md)

```
MEMORY_EXPANSION_QUAD_DENOMINATOR := 512
MEMORY_EXPANSION_LINEAR_COEFF := 3
calc_memory_cost(memory_size) := MEMORY_EXPANSION_LINEAR_COEFF * memory_size + floor(memory_size * memory_size / MEMORY_EXPANSION_QUAD_DENOMINATOR)

memory_expansion_gas_cost := calc_memory_cost(next.memory_size) - memory_cost(curr.memory_size)
```

From `ErrorOutOfGasMemoryCopy` md file, it says
```
    memory_expansion_gas_cost = (
        3 * (next_memory_word_size - current_memory_word_size)
        + next_memory_word_size * next_memory_word_size // 512
        - current_memory_word_size * current_memory_word_size // 512
    )

// this should be equal
(3 * next_memory_word_size + next_memory_word_size * next_memory_word_size // 512) 
- (3 * next_memory_word_size + current_memory_word_size * current_memory_word_size // 512)
==> 
calc_memory_cost(next_memory_word_size) - memory_cost(current_memory_word_size)
```

Correct me if I misunderstand anything.